### PR TITLE
Add metrics to measure latency of sub-operations in Controller createTransaction and commit Transaction APIs

### PR DESCRIPTION
**Pull request by pbelgundi**
Signed-off-by: pbelgundi <prajakta.belgundi@emc.com>

**Change log description**  
This PR has changes for measuring latency of operations inside create & commit Controller APIs as well as for commit event processing.

**Purpose of the change**  

**What the code does**  
Measure latency of Txn sub-operations and log it.

**How to verify it**  
All Unit and Integration Tests should pass

### CODE REVIEW

File: controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/CommitRequestHandler.java [hunk 1]
```diff
@@ -347,13 +355,21 @@ public CommitRequestHandler(final StreamMetadataStore streamMetadataStore,
...
+                    })
                     .thenRun(() -> TransactionMetrics.getInstance().commitTransaction(scope, stream, timer.getElapsed()));
         }
-        
+
         return future
                 .thenCompose(v -> bucketStore.addStreamToBucketStore(BucketStore.ServiceType.WatermarkingService, scope,
-                        stream, executor));
+                        stream, executor))
+                .thenAccept(v -> TransactionMetrics.getInstance().reportCommitTransactionBatchCount(scope, stream, transactionsToCommit.size()));
```
Comment (andreipaduroiu) [inline controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/CommitRequestHandler.java:372]: `thenRun`

File: controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/CommitRequestHandler.java [hunk 2]
```diff
@@ -347,13 +355,21 @@ public CommitRequestHandler(final StreamMetadataStore streamMetadataStore,
                     // Note: if its a rerun, transaction commit offsets may have been updated already in previous iteration
                     // so this will not update/modify it. 
                     .thenCompose(v -> streamMetadataTasks.notifyTxnCommit(scope, stream, segments, txnId, context.getRequestId()))
-                    .thenCompose(map -> streamMetadataStore.recordCommitOffsets(scope, stream, txnId, map, context, executor))
+                    .thenCompose(map -> {
+                        Timer recordOffsetsTimer = new Timer();
+                        return streamMetadataStore.recordCommitOffsets(scope, stream, txnId, map, context, executor)
+                                .thenApply(x -> {
```
Comment (andreipaduroiu) [inline controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/CommitRequestHandler.java:361]: `thenRun` - I don't see you're actually using `x` downstream

File: shared/metrics/src/main/java/io/pravega/shared/MetricsNames.java [hunk 3]
```diff
@@ -296,11 +296,25 @@
...
+    public static final String COMMITTING_TRANSACTION_REMOVE_TIMEOUT_SVC_LATENCY = PREFIX + "controller.transactions.committed_remove_timeout_svc_latency_ms";  // Histogram
+    public static final String COMMIT_TRANSACTION_START_LATENCY = PREFIX + "controller.transactions.committed_start_latency_ms";                                       // Counter and Per-stream Counter
+    public static final String COMMIT_TRANSACTION_RECORD_OFFSETS_LATENCY = PREFIX + "controller.transactions.committed_record_offsets_latency_ms";                    // Histogram
+    public static final String COMMIT_TRANSACTION_ROLLOVER_LATENCY = PREFIX + "controller.transactions.committed_rollover_latency_ms";                                       // Counter and Per-stream Counter
+    public static final String COMMIT_TRANSACTION_COMPLETE_LATENCY = PREFIX + "controller.transactions.committed_complete_latency_ms";                                       // Counter and Per-stream Counter
+    public static final String COMMIT_TRANSACTION_BATCH_COUNT = PREFIX + "controller.transactions.committed_batch_Count";                                       // Counter and Per-stream Counter
```
Comment (andreipaduroiu) [inline shared/metrics/src/main/java/io/pravega/shared/MetricsNames.java:316]: lowercase `C`

Comment (pbelgundi) [reply shared/metrics/src/main/java/io/pravega/shared/MetricsNames.java]: fixed

File: controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/CommitRequestHandler.java (hunk 2 above)
Comment (pbelgundi) [reply controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/CommitRequestHandler.java]: fixed

File: controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/CommitRequestHandler.java (hunk 1 above)
Comment (pbelgundi) [reply controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/CommitRequestHandler.java]: fixed

File: controller/src/main/java/io/pravega/controller/store/PravegaTablesStoreHelper.java [hunk 4]
```diff
@@ -357,7 +357,7 @@ public void invalidateCache(String table, String key) {
      */
     public <T> CompletableFuture<VersionedMetadata<T>> getEntry(String tableName, String key, Function<byte[], T> fromBytes,
                                                                 long requestId) {
-        log.trace(requestId, "get entry called for : {} key : {}", tableName, key);
+        log.info(requestId, "get entry called for : {} key : {}", tableName, key);
```
Comment (andreipaduroiu) [inline controller/src/main/java/io/pravega/controller/store/PravegaTablesStoreHelper.java:360]: Is INFO appropriate? Will this pollute logs? Maybe DEBUG is better if this is invoked frequently.

File: controller/src/test/java/io/pravega/controller/server/eventProcessor/ControllerEventProcessorTest.java [hunk 5]
```diff
@@ -387,7 +387,7 @@ private void checkTransactionState(String scope, String stream, UUID txnId, TxnS
         assertEquals(expectedStatus, txnStatus);
     }
 
-    @Test(timeout = 10000)
+    @Test//(timeout = 10000)
```
Comment (andreipaduroiu) [inline controller/src/test/java/io/pravega/controller/server/eventProcessor/ControllerEventProcessorTest.java:390]: Revert

File: controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/CommitRequestHandler.java [hunk 6]
```diff
@@ -159,81 +162,117 @@ public CommitRequestHandler(final StreamMetadataStore streamMetadataStore,
...
-                                    // reset state conditionally in case we were left with stale committing state from
-                                    // a previous execution
-                                    // that died just before updating the state back to ACTIVE but after having 
-                                    // completed all the work.
-                                    return CompletableFuture.completedFuture(versionedMetadata);
-                                } else {
-                                    int txnEpoch = versionedMetadata.getObject().getEpoch();
-                                    List<UUID> txnList = versionedMetadata.getObject().getTransactionsToCommit();
+                            streamMetadataStore.startCommitTransactions(scope, stream, MAX_TRANSACTION_COMMIT_BATCH_SIZE, context, executor)
+                                    .thenApply(commitTxnRecord -> {
```
Comment (andreipaduroiu) [inline controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/CommitRequestHandler.java:175]: You can merge this `thenApply` into the next `thenComposeAsync` - fewer callbacks will improve your app performance.

**Review by andreipaduroiu — COMMENTED** [review]
@pbelgundi  I left a few more comments. This PR looks old and has merge conflicts - are we still pursuing it?

**Comment by pbelgundi** [comment]
@andreipaduroiu This PR was raised for transactions performance testing and we didn't plan to merge it to master.
Closing it now, as the changes have already been moved to a test branch `feature-txns-perf-test`

**Pull request closed without merging**